### PR TITLE
Add package evaluation fuzzer

### PR DESCRIPTION
**Pull request by instagibbs**
This fuzzer target caught the issue in https://github.com/bitcoin/bitcoin/pull/28251 within 5 minutes on master branch, and an additional issue which I've applied a preliminary patch to cover.

Fuzzer target does the following:

1) Picks mempool confgs, including max package size, count, mempool size, etc
2) Generates 1 to 26 transactions with arbitrary coins/fees, the first N-1 spending only confirmed outpoints
3) Nth transaction, if >1, sweeps all unconfirmed outpoints in mempool
4) If N==1, it may submit it through single-tx submission path, to allow for more interesting topologies
5) Otherwise submits through package submission interface
6) Repeat 1-5  a few hundred times per mempool instance

In other words, it ends up building chains of txns in the mempool using parents-and-children packages, which is currently the topology supported on master.

The test itself is a direct rip of tx_pool.cpp, with a number of assertions removed because they were failing for unknown reasons, likely due to the notification changes of single tx submission to package, which is used to track addition/removal of transactions in the test. I'll continue working on re-adding these assertions for further invariant testing.

### CODE REVIEW

**Comment by instagibbs** [comment]
cc @glozow @MarcoFalke

File: src/test/fuzz/package_eval.cpp [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+    args.ForceSetArg("-limitancestorsize",
+                     ToString(fuzzed_data_provider.ConsumeIntegralInRange<unsigned>(0, 202)));
+    args.ForceSetArg("-limitdescendantcount",
+                     ToString(fuzzed_data_provider.ConsumeIntegralInRange<unsigned>(0, 50)));
+    args.ForceSetArg("-limitdescendantsize",
+                     ToString(fuzzed_data_provider.ConsumeIntegralInRange<unsigned>(0, 202)));
+    args.ForceSetArg("-maxmempool",
+                     ToString(fuzzed_data_provider.ConsumeIntegralInRange<unsigned>(0, 200)));
+    args.ForceSetArg("-mempoolexpiry",
+                     ToString(fuzzed_data_provider.ConsumeIntegralInRange<unsigned>(0, 999)));
```
Comment (glozow) [inline src/test/fuzz/package_eval.cpp:88]: Maybe worth adding `-bytespersigop` too?

**Comment by darosior** [comment]
Awesome, Concept ACK.

**Review by glozow — COMMENTED** [review]
concept ACK, really nice :rocket:

File: src/test/fuzz/package_eval.cpp (hunk 1 above)
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:88]: to take advantage of this, probably would need to do some bare mulitisigs? wouldn't hurt to add I suppose

**Review by dergoegge — COMMENTED** [review]
Concept ACK

**Review by brunoerg — COMMENTED** [review]
Concept ACK

File: src/test/fuzz/package_eval.cpp [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+void MockTime(FuzzedDataProvider& fuzzed_data_provider, const Chainstate& chainstate)
+{
+    const auto time = ConsumeTime(fuzzed_data_provider,
+                                  chainstate.m_chain.Tip()->GetMedianTimePast() + 1,
+                                  std::numeric_limits<decltype(chainstate.m_chain.Tip()->nTime)>::max());
+    SetMockTime(time);
+}
+
+CTxMemPool MakeMempool(FuzzedDataProvider& fuzzed_data_provider, const NodeContext& node)
```
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:124]: In ddc6f41a1c72560b1d7701db921c0efb280c20b8: Is this `MakeMempool` the same from `tx_pool`? If so, perhaps we could move it to a util?

Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:124]: In ddc6f41a1c72560b1d7701db921c0efb280c20b8: It seems to have some functions from `tx_pool`. Couldn't we have them in a `util` instead of duplicating them?

File: src/test/fuzz/package_eval.cpp [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+{
+    // Take the default options for tests...
+    CTxMemPool::Options mempool_opts{MemPoolOptionsForTest(node)};
+
+    // ...override specific options for this specific fuzz suite
+    mempool_opts.estimator = nullptr;
+    mempool_opts.check_ratio = 1;
+    mempool_opts.require_standard = fuzzed_data_provider.ConsumeBool();
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:123]: done

**Comment by instagibbs** [comment]
rebased on master, removed arg round-trip, un-marked WIP

**Comment by glozow** [comment]
Any interest in using `CheckPackageMempoolAcceptResult` from https://github.com/bitcoin/bitcoin/pull/26711/commits/baf475de599c82e7718ffefd7f0481fc88217853? It checks that a result contains the fields we expect based on whether it's supposed to be valid.

File: src/test/fuzz/package_eval.cpp [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+                    in.scriptWitness.stack = script_wit_stack;
+
+                    tx_mut.vin.push_back(in);
+                }
+                const auto amount_fee = fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(0, amount_in);
+                const auto amount_out = (amount_in - amount_fee) / num_out;
+                for (int i = 0; i < num_out; ++i) {
+                    tx_mut.vout.emplace_back(amount_out, P2WSH_OP_TRUE);
+                }
+                // TODO vary transaction sizes to catch size-related issues
```
Comment (instagibbs) [inline src/test/fuzz/package_eval.cpp:211]: making the tx >40kWu sometimes would have likely allowed the fuzzer to catch https://github.com/bitcoin/bitcoin/pull/28472

**Comment by instagibbs** [comment]
removed the second commit to not impede review of the various bugfix PRs individually

https://github.com/bitcoin/bitcoin/pull/28471
https://github.com/bitcoin/bitcoin/pull/28472

File: src/test/fuzz/package_eval.cpp [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+    const auto& node = g_setup->m_node;
+    auto& chainstate{static_cast<DummyChainState&>(node.chainman->ActiveChainstate())};
+
+    MockTime(fuzzed_data_provider, chainstate);
+
+    // All RBF-spendable outpoints outside of the immediate package
+    std::set<COutPoint> outpoints_rbf;
+    // All outpoints counting toward the total supply (subset of outpoints_rbf)
+    std::set<COutPoint> outpoints_supply;
```
Comment (achow101) [inline src/test/fuzz/package_eval.cpp:140]: I don't think this set is ever actually used.

File: src/test/fuzz/package_eval.cpp [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+            // This is empty if it fails early checks, or "full" if transactions are looked at deeper
+            Assert(result_package.m_tx_results.size() == txs.size() || result_package.m_tx_results.empty());
+            if (result_package.m_state.GetResult() == PackageValidationResult::PCKG_POLICY) {
+                for (const auto& tx : txs) {
+                    removed.erase(tx);
+                }
+            } else {
+                for (const auto& [k, v] : result_package.m_tx_results) {
+                    if (v.m_result_type != MempoolAcceptResult::ResultType::INVALID) {
+                        Assert(true);
```
Comment (achow101) [inline src/test/fuzz/package_eval.cpp:295]: `Assert(true)` seems to not be useful.

File: src/test/fuzz/package_eval.cpp [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+CTxMemPool MakeMempool(FuzzedDataProvider& fuzzed_data_provider, const NodeContext& node)
+{
+    // Take the default options for tests...
+    CTxMemPool::Options mempool_opts{MemPoolOptionsForTest(node)};
+
+
+    // ...override specific options for this specific fuzz suite
+    mempool_opts.limits.ancestor_count = fuzzed_data_provider.ConsumeIntegralInRange<unsigned>(0, 50);
+    mempool_opts.limits.ancestor_size_vbytes = fuzzed_data_provider.ConsumeIntegralInRange<unsigned>(0, 202) * 1'000;
+    mempool_opts.limits.descendant_count = fuzzed_data_provider.ConsumeIntegralInRange<unsigned>(0, 50);
```
Comment (achow101) [inline src/test/fuzz/package_eval.cpp:115]: Seems a bit odd to me that the chain limits could go up to 50 but we only create packages of up to 26.

Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:115]: That's chain limits vs package limit checks, I think it's useful

File: src/test/fuzz/package_eval.cpp (hunk 5 above)
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:140]: it is added to in `insert_tx` when used in `created_by_tx` arg. Lets each new package consider newly-entered outpoints in the mempool

File: src/test/fuzz/package_eval.cpp (hunk 6 above)
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: removed debugging code

**Comment by instagibbs** [comment]
updated with cleanups

**Comment by instagibbs** [comment]
from discussions with others:

1) I have a PR built on https://github.com/bitcoin/bitcoin/pull/26711 which generalized from this fairly basic topology to general ancestor packages, including the last tx spending any subset of outpoints, and any package transactions potentially spending other package outpoints. I'd like to make sure it's getting relatively targeted coverage before switching to that, so deferring to after this PR.
2) We can add coverage where we mutate the wtxid of txns
3) Also should allow duplication of inputs

I plan on doing these in the follow-up on top of #26711

File: src/test/fuzz/package_eval.cpp [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+        RegisterSharedValidationInterface(txr);
+        const bool bypass_limits = fuzzed_data_provider.ConsumeBool();
+
+        // Single-tx packages should be rejected, so do that sometimes, and sometimes send it via single submission
+        // to allow it into the mempool by itself to make more interesting mempool packages
+        auto single_submit = txs.size() == 1 && fuzzed_data_provider.ConsumeBool();
+        auto package_submit = !single_submit;
+
+        const auto result_package = WITH_LOCK(::cs_main,
+                                    return ProcessNewPackage(chainstate, tx_pool, txs, /*test_accept=*/!package_submit));
```
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:257]: In bf7436cde8ba7eac0912817f6f8a41828cb60bc2: In `ProcessNewPackage`, couldn't we just use `single_submit` for `/*test_accept=*/` directly, instead of creating `package_submit`?

File: src/test/fuzz/package_eval.cpp [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+
+    // Seeded by coinbase outputs first
+    outpoints_rbf = outpoints_supply;
+
+    CTxMemPool tx_pool_{MakeMempool(fuzzed_data_provider, node)};
+    MockedTxPool& tx_pool = *static_cast<MockedTxPool*>(&tx_pool_);
+
+    chainstate.SetMempool(&tx_pool);
+
+    const CCoinsViewMemPool amount_view{WITH_LOCK(::cs_main, return &chainstate.CoinsTip()), tx_pool};
```
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:155]: In bf7436cde8ba7eac0912817f6f8a41828cb60bc2: I think `amount_view` hasn't been used.

Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: deleted

File: src/test/fuzz/package_eval.cpp (hunk 8 above)
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:256]: I find this ever so slightly more readable myself, opinions may vary

**Comment by instagibbs** [comment]
rebased on latest master on top of #28471 

Should not be crashing anymore :crossed_fingers: cc @dergoegge

File: src/test/fuzz/package_eval.cpp [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+            // Note that this test currently only spends unconfirmed outputs in last transaction.
+            bool last_tx = num_txs > 1 && txs.size() == num_txs - 1;
+
+            // Create transaction to add to the mempool
+            const CTransactionRef tx = [&] {
+                CMutableTransaction tx_mut;
+                tx_mut.nVersion = CTransaction::CURRENT_VERSION;
+                tx_mut.nLockTime = fuzzed_data_provider.ConsumeBool() ? 0 : fuzzed_data_provider.ConsumeIntegral<uint32_t>();
+                // Last tx will sweep all outpoints in package
+                const auto num_in = last_tx ? package_outpoints.size()  : fuzzed_data_provider.ConsumeIntegralInRange<int>(1, outpoints_rbf.size());
```
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:179]:
```suggestion
                const auto num_in = last_tx ? package_outpoints.size() : fuzzed_data_provider.ConsumeIntegralInRange<int>(1, outpoints_rbf.size());
```

File: src/test/fuzz/package_eval.cpp [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+                    outpoints_value[COutPoint(tx->GetHash(), i)] = tx->vout[i].nValue;
+                }
+                return tx;
+            }();
+            txs.push_back(tx);
+            wtxid_to_tx[tx->GetWitnessHash()] = tx;
+        }
+
+        if (fuzzed_data_provider.ConsumeBool()) {
+            MockTime(fuzzed_data_provider, chainstate);
```
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:230]: Why is it calling `MockTime` again? I guess it's to trigger the `nLockTime`?

File: src/test/fuzz/package_eval.cpp [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+        }
+        if (fuzzed_data_provider.ConsumeBool()) {
+            tx_pool.RollingFeeUpdate();
+        }
+        if (fuzzed_data_provider.ConsumeBool()) {
+            const auto& txid = fuzzed_data_provider.ConsumeBool() ?
+                                   txs.back()->GetHash() :
+                                   PickValue(fuzzed_data_provider, outpoints_rbf).hash;
+            const auto delta = fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(-50 * COIN, +50 * COIN);
+            tx_pool.PrioritiseTransaction(txid, delta);
```
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:240]: In 2ecdc2183f0a1291730ae8728fa3d847b9cf80b5: Couldn't we prioritise more than 1 transaction?

File: src/test/fuzz/package_eval.cpp [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate, txs.back(), GetTime(), bypass_limits, /*test_accept=*/!single_submit));
+        const bool accepted = res.m_result_type == MempoolAcceptResult::ResultType::VALID;
+
+        SyncWithValidationInterfaceQueue();
+        UnregisterSharedValidationInterface(txr);
+
+        if (single_submit) {
+            Assert(accepted != added.empty());
+            Assert(accepted == res.m_state.IsValid());
+            Assert(accepted != res.m_state.IsInvalid());
```
Comment (brunoerg) [inline src/test/fuzz/package_eval.cpp:276]: In `2ecdc2183f0a1291730ae8728fa3d847b9cf80b5`: Does it make sense to check whether the state is invalid if we previously assert that is valid?

File: src/test/fuzz/package_eval.cpp (hunk 11 above)
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:230]: it's called for each package separately, sometimes, yes

File: src/test/fuzz/package_eval.cpp (hunk 12 above)
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:240]: yes, I'm not sure if it will end up covering more interesting cases though

File: src/test/fuzz/package_eval.cpp (hunk 13 above)
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:276]: sorry can you be a bit more explicit in what you're asking to do?

File: src/test/fuzz/package_eval.cpp [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+        std::set<CTransactionRef> removed;
+        std::set<CTransactionRef> added;
+        auto txr = std::make_shared<TransactionsDelta>(removed, added);
+        RegisterSharedValidationInterface(txr);
+        const bool bypass_limits = fuzzed_data_provider.ConsumeBool();
+
+        // Single-tx packages should be rejected, so do that sometimes, and sometimes send it via single submission
+        // to allow it into the mempool by itself to make more interesting mempool packages
+        auto single_submit = txs.size() == 1 && fuzzed_data_provider.ConsumeBool();
+        auto package_submit = !single_submit;
```
[the remark this answers is not in the document]
Comment (glozow) [reply src/test/fuzz/package_eval.cpp:253]: Also... wondering why allow `num_txs` to be less than 2 anyway? Isn't this test kind of the same as what's in the tx_pool fuzzer?

**Review by glozow — COMMENTED** [review]
A few quick questions, looks correct to me though some bits are a little confusing. I know future updates are planned so feel free to save for later if I'm scope creeping too much.

**Comment by glozow** [comment]
CI failure is a functional test so I'd assume unrelated

**Comment by dergoegge** [comment]
Made a coverage report: https://dergoegge.github.io/bitcoin-coverage/pr28450/fuzz.coverage/index.html

**Comment by glozow** [comment]
> Made a coverage report: https://dergoegge.github.io/bitcoin-coverage/pr28450/fuzz.coverage/index.html

Looks pretty good to me, noting the main missing bits:
- same-txid-different-witness
- policy script checks failing
- bad-txns-too-many-sigops
- being granted CPFP carve out and then failing

^which have already been marked as followups.

There's no package-not-child-with-unconfirmed-parents which is by design. Also looks like RBF rules are red, perhaps it'll take a while to hit them?

**Comment by dergoegge** [comment]
> Also looks like RBF rules are red, perhaps it'll take a while to hit them?

The seed corpus from the report was the result of 5000+ cpu hours, so I'd say it's likely just not able to hit the rbf paths.

File: src/test/fuzz/package_eval.cpp [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+void Finish(FuzzedDataProvider& fuzzed_data_provider, MockedTxPool& tx_pool, Chainstate& chainstate)
+{
+    WITH_LOCK(::cs_main, tx_pool.check(chainstate.CoinsTip(), chainstate.m_chain.Height() + 1));
+    {
+        BlockAssembler::Options options;
+        options.nBlockMaxWeight = fuzzed_data_provider.ConsumeIntegralInRange(0U, MAX_BLOCK_WEIGHT);
+        options.blockMinFeeRate = CFeeRate{ConsumeMoney(fuzzed_data_provider, /*max=*/COIN)};
+        auto assembler = BlockAssembler{chainstate, &tx_pool, options};
+        auto block_template = assembler.CreateNewBlock(CScript{} << OP_TRUE);
+        Assert(block_template->block.vtx.size() >= 1);
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: removed as it's not very interesting a check

File: src/test/fuzz/package_eval.cpp [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+    }
+    const auto info_all = tx_pool.infoAll();
+    if (!info_all.empty()) {
+        const auto& tx_to_remove = *PickValue(fuzzed_data_provider, info_all).tx;
+        WITH_LOCK(tx_pool.cs, tx_pool.removeRecursive(tx_to_remove, MemPoolRemovalReason::BLOCK /* dummy */));
+        std::vector<uint256> all_txids;
+        tx_pool.queryHashes(all_txids);
+        assert(all_txids.size() < info_all.size());
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: done

File: src/test/fuzz/package_eval.cpp [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+        std::set<COutPoint> package_outpoints;
+        while (txs.size() < num_txs) {
+
+            // Last transaction in a package needs to be a child of parents to get further in validation
+            // so the last transaction to be generated(in a >1 package) must spend all package-made outputs
+            // Note that this test currently only spends unconfirmed outputs in last transaction.
+            bool last_tx = num_txs > 1 && txs.size() == num_txs - 1;
+
+            // Create transaction to add to the mempool
+            const CTransactionRef tx = [&] {
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:203]: done

File: src/test/fuzz/package_eval.cpp [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+                    tx_mut.vin.push_back(in);
+                }
+                const auto amount_fee = fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(0, amount_in);
+                const auto amount_out = (amount_in - amount_fee) / num_out;
+                for (int i = 0; i < num_out; ++i) {
+                    tx_mut.vout.emplace_back(amount_out, P2WSH_OP_TRUE);
+                }
+                // TODO vary transaction sizes to catch size-related issues
+                auto tx = MakeTransactionRef(tx_mut);
+                // Restore previously removed outpoints, except in-package outpoints
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:244]: yes

File: src/test/fuzz/package_eval.cpp [hunk 19]
```diff
@@ -0,0 +1,338 @@
...
+                   it->second.m_result_type == MempoolAcceptResult::ResultType::MEMPOOL_ENTRY);
+        }
+
+        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate, txs.back(), GetTime(), bypass_limits, /*test_accept=*/!single_submit));
+        const bool accepted = res.m_result_type == MempoolAcceptResult::ResultType::VALID;
+
+        SyncWithValidationInterfaceQueue();
+        UnregisterSharedValidationInterface(txr);
+
+        if (single_submit) {
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:310]: done

File: src/test/fuzz/package_eval.cpp [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+                if (!last_tx) {
+                    for (const auto& in : tx->vin) {
+                        Assert(outpoints.insert(in.prevout).second);
+                    }
+                }
+                // Cache the in-package outpoints being made
+                for (size_t i = 0; i < tx->vout.size(); ++i) {
+                    package_outpoints.emplace(tx->GetHash(), i);
+                    outpoints_value[COutPoint(tx->GetHash(), i)] = tx->vout[i].nValue;
+                }
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: done, had to keep the outpoints_value check outside though since it needs all generated outpoints

File: src/test/fuzz/package_eval.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: > Also... wondering why allow num_txs to be less than 2 anyway?

1) it can get rejected by ProcessNewPackage
2) it can cause the existing mempool to be somewhat more interesting maybe

**Comment by instagibbs** [comment]
> Also looks like RBF rules are red, perhaps it'll take a while to hit them?

A couple of those cases would look difficult to hit without intention/design to do so. One I'm unsure how it hasn't been hit, something to look at later.

> I think achow is wondering if this is redundant with outpoints_rbf? Since new transactions are always created using outpoints_rbf inputs.

Can't find the actual place to reply(thanks github), removed outpoints_supply entirely as it's vestigial

----

Took all suggestions, pushed

File: src/test/fuzz/package_eval.cpp [hunk 21]
```diff
@@ -0,0 +1,341 @@
...
+#include <validationinterface.h>
+
+using node::BlockAssembler;
+using node::NodeContext;
+
+namespace {
+
+const TestingSetup* g_setup;
+std::vector<COutPoint> g_outpoints_coinbase_init_mature;
+std::vector<COutPoint> g_outpoints_coinbase_init_immature;
```
Comment (glozow) [inline src/test/fuzz/package_eval.cpp:28]: note: `g_outpoints_coinbase_init_immature` seems unused rn, I assume the plan is to maybe spend immature coinbases later?

File: src/test/fuzz/package_eval.cpp [hunk 22]
```diff
@@ -0,0 +1,341 @@
...
+
+
+        // Make packages of 1-to-26 transactions
+        const auto num_txs = (size_t) fuzzed_data_provider.ConsumeIntegralInRange<int>(1, 26);
+        std::set<COutPoint> package_outpoints;
+        while (txs.size() < num_txs) {
+
+            // Last transaction in a package needs to be a child of parents to get further in validation
+            // so the last transaction to be generated(in a >1 package) must spend all package-made outputs
+            // Note that this test currently only spends unconfirmed outputs in last transaction.
```
Comment (glozow) [inline src/test/fuzz/package_eval.cpp:199]: nit: `outpoints_rbf` contains unconfirmed, no?
```suggestion
            // Note that this test currently only spends package outputs in last transaction.
```

File: src/test/fuzz/package_eval.cpp [hunk 23]
```diff
@@ -0,0 +1,341 @@
...
+        // There is only 1 transaction in the package. We did a test-package-accept and a ATMP
+        if (single_submit) {
+            Assert(accepted != added.empty());
+            Assert(accepted == res.m_state.IsValid());
+            if (accepted) {
+                Assert(added.size() == 1);
+                Assert(txs.back() == *added.begin());
+            } else {
+                // Do not consider rejected transaction removed
+                removed.erase(txs.back());
```
Comment (glozow) [inline src/test/fuzz/package_eval.cpp:318]: nit: I think these `removed.erase()` calls might be unnecessary now that you don't use it to update `outpoints_rbf`.

**Review by glozow — COMMENTED** [review]
ACK 3707ed7c0d02d86ff86c159bbd746acfa52db145
Code and coverage look good. Will run overnight to sanity check that outpointsupdater hasn't broken anything.

Couple more ideas for post-26711 expansions:
- add nonexistent outpoint to hit missing inputs cases
- shuffle before submission

File: src/test/fuzz/package_eval.cpp [hunk 24]
```diff
@@ -0,0 +1,341 @@
...
+        if (fuzzed_data_provider.ConsumeBool()) {
+            const auto& txid = fuzzed_data_provider.ConsumeBool() ?
+                                   txs.back()->GetHash() :
+                                   PickValue(fuzzed_data_provider, outpoints_rbf).hash;
+            const auto delta = fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(-50 * COIN, +50 * COIN);
+            tx_pool.PrioritiseTransaction(txid, delta);
+        }
+
+        // Remember all removed and added transactions
+        std::set<CTransactionRef> removed;
```
Comment (achow101) [inline src/test/fuzz/package_eval.cpp:279]: `removed` doesn't seem to actually be used anywhere.

File: src/test/fuzz/package_eval.cpp (hunk 22 above)
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: taken

File: src/test/fuzz/package_eval.cpp (hunk 23 above)
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: removed `removed` entirely

File: src/test/fuzz/package_eval.cpp (hunk 24 above)
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: removed `removed` entirely

**Comment by instagibbs** [comment]
Removed the `Final` subroutine to get a slight speedup since it seems unnecessary, leaving just the `tx_pool.check()` which should catch all issues.

**Comment by instagibbs** [comment]
Going to hold off on all subsequent nits to get this in to un-draft #26711 and start building on top

**Comment by instagibbs** [comment]
https://github.com/bitcoin/bitcoin/pull/28450/checks?check_run_id=17181111860 multiprocess build having an issue of some kind

File: src/test/fuzz/package_eval.cpp [hunk 25]
```diff
@@ -0,0 +1,297 @@
...
+
+    void TransactionAddedToMempool(const CTransactionRef& tx, uint64_t /* mempool_sequence */) override
+    {
+        // Transactions may be entered and booted any number of times
+        m_added.insert(tx);
+    }
+
+    void TransactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRemovalReason reason, uint64_t /* mempool_sequence */) override
+    {
+        // Transactions may be entered and booted any number of times
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:96]: Sounds right, surprised I haven't hit this since it quite often chooses tiny mempool sizes too small for even a single transaction to enter.

Taken.

File: src/test/fuzz/package_eval.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: removed this already circa https://github.com/bitcoin/bitcoin/pull/28450/commits/70879e4f02e522ee82b563917ef6700efe4cf5dd

File: src/test/fuzz/package_eval.cpp [hunk 26]
```diff
@@ -0,0 +1,298 @@
...
+FUZZ_TARGET(tx_package_eval, .init = initialize_tx_pool)
+{
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+    const auto& node = g_setup->m_node;
+    auto& chainstate{static_cast<DummyChainState&>(node.chainman->ActiveChainstate())};
+
+    MockTime(fuzzed_data_provider, chainstate);
+
+    // All RBF-spendable outpoints outside of the unsubmitted package
+    std::set<COutPoint> outpoints_rbf;
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:141]: was going to change this anyways, renaming to `mempool_outpoints`

File: src/test/fuzz/package_eval.cpp [hunk 27]
```diff
@@ -0,0 +1,298 @@
...
+    MockedTxPool& tx_pool = *static_cast<MockedTxPool*>(&tx_pool_);
+
+    chainstate.SetMempool(&tx_pool);
+
+    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 300)
+    {
+        Assert(!outpoints_rbf.empty());
+
+        std::vector<CTransactionRef> txs;
+        std::map<uint256, CTransactionRef> wtxid_to_tx;
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp:161]: removed

File: src/test/fuzz/package_eval.cpp (hunk 8 above)
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: ok fine, removed :)

**Comment by murchandamus** [comment]
ACK 262ab8ef7860d43cebc9d04721e3a075b4edf06e

Note that this is the first time I review something from the package relay project, so take that for whatever it’s worth

**Review by glozow — COMMENTED** [review]
reACK 262ab8ef7860d43cebc9d04721e3a075b4edf06e

**Review by dergoegge — APPROVED** [review]
tACK 262ab8ef7860d43cebc9d04721e3a075b4edf06e

Will generate and submit inputs to qa-assets once this is merged

File: src/test/fuzz/package_eval.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/test/fuzz/package_eval.cpp]: Oh I was mistaken, AcceptPackage takes package sizes of 1(I was thinking of BIP331 probably!). This PR could have been a bit simpler I think.
